File: oracle/tokenize_abi.py
from __future__ import annotations

from typing import Any, Dict, Iterable, Optional


class TokenizeError(RuntimeError):
    def __init__(self, code: str, msg: str, line: int = 0, col: int = 0) -> None:
        super().__init__(msg)
        self.code = code
        self.msg = msg
        self.line = line
        self.col = col


def abi_tokenize_ok(text: str, tokenizer_abi: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    """Validate text against a tokenizer ABI.

    Returns None when valid, otherwise an error dict with code/msg/line/col.
    """
    allowed_ranges = tokenizer_abi.get("allowed_codepoint_ranges") or []
    forbidden_ranges = tokenizer_abi.get("forbidden_codepoint_ranges") or []
    allowed_set = set(tokenizer_abi.get("allowed_codepoints") or [])
    forbidden_set = set(tokenizer_abi.get("forbidden_codepoints") or [])

    for idx, ch in enumerate(text):
        cp = ord(ch)
        line, col = _line_col(text, idx)

        if cp in forbidden_set or _in_ranges(cp, forbidden_ranges):
            return {
                "code": "E_TOK_FORBIDDEN",
                "msg": f"forbidden codepoint: U+{cp:04X}",
                "line": line,
                "col": col,
            }
        if allowed_set or allowed_ranges:
            if cp not in allowed_set and not _in_ranges(cp, allowed_ranges):
                return {
                    "code": "E_TOK_DISALLOWED",
                    "msg": f"disallowed codepoint: U+{cp:04X}",
                    "line": line,
                    "col": col,
                }

    return None


def _in_ranges(cp: int, ranges: Iterable[Iterable[int]]) -> bool:
    for pair in ranges:
        if len(pair) != 2:
            continue
        start, end = pair
        if start <= cp <= end:
            return True
    return False


def _line_col(text: str, idx: int) -> tuple[int, int]:
    line = text.count("\n", 0, idx) + 1
    last_nl = text.rfind("\n", 0, idx)
    col = idx + 1 if last_nl == -1 else idx - last_nl
    return line, col

File: oracle-js/tokenize_abi.js
export function abiTokenizeOk(text, tokenizerAbi) {
  const allowedRanges = tokenizerAbi.allowed_codepoint_ranges || [];
  const forbiddenRanges = tokenizerAbi.forbidden_codepoint_ranges || [];
  const allowedSet = new Set(tokenizerAbi.allowed_codepoints || []);
  const forbiddenSet = new Set(tokenizerAbi.forbidden_codepoints || []);

  for (let idx = 0; idx < text.length; idx += 1) {
    const cp = text.codePointAt(idx);
    const { line, col } = lineCol(text, idx);

    if (forbiddenSet.has(cp) || inRanges(cp, forbiddenRanges)) {
      return {
        code: "E_TOK_FORBIDDEN",
        msg: `forbidden codepoint: U+${cp.toString(16).toUpperCase().padStart(4, "0")}`,
        line,
        col,
      };
    }
    if (allowedSet.size || allowedRanges.length) {
      if (!allowedSet.has(cp) && !inRanges(cp, allowedRanges)) {
        return {
          code: "E_TOK_DISALLOWED",
          msg: `disallowed codepoint: U+${cp.toString(16).toUpperCase().padStart(4, "0")}`,
          line,
          col,
        };
      }
    }
    if (cp > 0xffff) idx += 1;
  }

  return null;
}

function inRanges(cp, ranges) {
  for (const pair of ranges) {
    if (!Array.isArray(pair) || pair.length !== 2) continue;
    const [start, end] = pair;
    if (start <= cp && cp <= end) return true;
  }
  return false;
}

function lineCol(text, idx) {
  const before = text.slice(0, idx);
  const line = before.split("\n").length;
  const lastNl = before.lastIndexOf("\n");
  const col = lastNl === -1 ? idx + 1 : idx - lastNl;
  return { line, col };
}
